File: 171044046_HW8 (java)/171044046/GTUVector.java
/**
 * GTUVector class
 *
 * @author Esra Eryılmaz
 * @version 1.0
 * @since 2020-01-16
 */

import java.lang.IllegalArgumentException;


public class GTUVector<T> extends GTUContainer<T>
{
	/**
	 *returns true if the vector is empty
	 */
	@Override
	public boolean empty()
	{
		if(size == 0)
			return true;
		else
			return false;
	}

	/**
	 *@return size of the vector
	 */
	@Override
	public int size()
	{
		return size;
	}

	/**
	 *@return maximum size of the vector
	 */
	@Override
	public int max_size()
	{
		return size;	//for size I keep just one data and I return that size;
	}

	/**
	 *@param val value for insertion
	 */
	@Override
	public void insert(T val) throws IllegalArgumentException
	{
		@SuppressWarnings("unchecked")
		T[] temp = (T[])new Object[size+1]; 

		for(int i=0 ; i<size ; ++i) // copy the previous.
			temp[i] = data[i];

		size=size+1;
		temp[size-1] = val; // add the new.
		data = temp;
	}

	/**
	 *deletes element from the vector
	 *@param val value for deletion
	 */
	@Override
	public void erase(T val)
	{
		int new_size=0;
		T [] temp = data;

		for(int i=0,j=0; i<size ; ++i, ++j)
		{
			if(data[i] != val)
			{
				temp[j] = data[i];
				++new_size;
			}
			else
				--j;
		}

		size = new_size;

		for(int i=0; i<size; ++i)
			data[i] = temp[i];
	}

	/**
	 *clear all content
	 */
	@Override
	public void clear()
	{
		data = null;
		size = 0;
	}

	/**
	 *returns iterator to beginning
	 */
	@Override
	public GTUIterator iterator()
	{
		GTUIterator it = new GTUIterator ();
		return it;
	}

	/**
	 *returns true if the vector contains the specified element
	 *@return true or false according to the vector contains
	 */
	@Override
	public boolean contains(Object o)
	{
		boolean exist = false;

		if(empty())
		{
			exist = false;
		}
		else
		{
			for(int i=0 ; i<size ; ++i)
			{
				if(data[i] == o)
					exist = true;
			}
		}

		return exist;
	}

}

File: 171044046_HW6/171044046/GTUVector.cpp
// GTUVector.cpp

#include "GTUVector.h"

namespace eryilmaz
{
	template<typename T>
	GTUVector<T>::GTUVector() :  GTUContainer<T>::GTUContainer()
	{
		//deliberately empty
	}

	template<typename T>
	GTUVector<T>::GTUVector(int inputSize) :  GTUContainer<T>::GTUContainer(inputSize)
	{
		//deliberately empty
	}

	template<typename T>
	GTUVector<T>::GTUVector(const GTUVector& other) :  GTUContainer<T>::GTUContainer(other)
	{
		//deliberately empty
	}

	template<typename T>
	bool GTUVector<T>::empty()const
	{
		if(this->_size == 0)
			return true;
		else
			return false;
	}

	template<typename T>
	int GTUVector<T>::size()const
	{
		return (this->_size);
	}

	template<typename T>
	int GTUVector<T>::max_size()const
	{
		return (this->_size);	//for size I keep just one data and I return that size;
	}

	template<typename T>
	void GTUVector<T>::insert(const T& val)
	{
		shared_ptr<T> temp = this->data;

		shared_ptr<T> temp2(new T[this->_size+1], default_delete<T[]>());
		this->data = temp2;

		for(int i=0 ; i<this->_size ; ++i)
			this->data.get()[i] = temp.get()[i];

		this->data.get()[this->_size] = val;
		this->_size +=1;
	}

	template<typename T>
	void GTUVector<T>::erase(const T& val)
	{
		int j=0;
		T* tmp = new T[this->_size];

		for(int i=0 ; i<this->_size ; ++i)
		{
			if(val != this->data.get()[i])
			{
				tmp[j] = this->data.get()[i];
				++j;
			}
		}

		this->_size -= 1;
		shared_ptr<T> temp(new T[this->_size], default_delete<T[]>());
		this->data = temp;

		for(int i=0 ; i<this->_size ; ++i)
			this->data.get()[i] = tmp[i];
		cout<< "Erase -> "<< val <<endl;
	}

	template<typename T>
	void GTUVector<T>::clear()
	{
		this->data = nullptr;
		this->_size = 0;
	}

	template<typename T>
	GTUIterator<T> GTUVector<T>::begin()const
	{
		return (this->data).get();
	}

	template<typename T>
	GTUIterator<T> GTUVector<T>::end()const
	{
		return (this->data).get() + (this->_size);
	}

	template<typename T>
	T& GTUVector<T>::operator[](const T& index)
	{
		return this->data.get()[index];
	}

}